src/stdlib_str2num.fypp

```diff
@@ -0,0 +1,553 @@
+    integer(kind=ikind), parameter :: BD         = ichar('D',kind=ikind) - digit_0
+    integer(kind=ikind), parameter :: LF = 10, CR = 13, WS = 32
+
+    interface to_num
+        #:for k1, t1 in INT_KINDS_TYPES
+        module procedure to_${k1}$
+        #:endfor
+        module procedure to_float
+        module procedure to_double
+#:if WITH_QP
+        module procedure to_quad
+#:endif
+    end interface
+
+    interface to_num_p
+        #:for k1, t1 in INT_KINDS_TYPES
+        module procedure to_${k1}$_p
+        #:endfor
+        module procedure to_float_p
+        module procedure to_double_p
+#:if WITH_QP
+        module procedure to_quad_p
+#:endif
+    end interface
+
+    interface to_num_base
+        #:for k1, t1 in INT_KINDS_TYPES
+        module procedure to_int_${k1}$
+        #:endfor
+        module procedure to_real_sp
+        module procedure to_real_dp
+#:if WITH_QP
+        module procedure to_real_qp
+#:endif
+    end interface
+    
+    contains
+    
+    !---------------------------------------------
+    ! String To Number interfaces
+    !---------------------------------------------
+
+    #:for k1, t1 in INT_KINDS_TYPES
+    elemental function to_${k1}$(s,mold) result(v)
+        ! -- In/out Variables
+        character(*), intent(in) :: s !> input string
+        ${t1}$, intent(in) :: mold !> dummy argument to disambiguate at compile time the generic interface
+        ${t1}$ :: v !> Output integer 32 value
+        ! -- Internal Variables
+        integer(int8) :: p !> position within the number
+        integer(int8)  :: stat !> error status
+        !----------------------------------------------
+        call to_num_base(s,v,p,stat)
+    end function
+    
+    function to_${k1}$_p(s,mold,stat) result(v)
+        ! -- In/out Variables
+        character(len=:), pointer :: s !> input string
+        ${t1}$, intent(in) :: mold !> dummy argument to disambiguate at compile time the generic interface
+        ${t1}$ :: v !> Output ${t1}$ value
+        integer(int8),intent(inout), optional :: stat
+        ! -- Internal Variables
+        integer(int8) :: p !> position within the number
+        integer(int8) :: err
+        !----------------------------------------------
+        call to_num_base(s,v,p,err)
+        p = min( p , len(s) )
+        s => s(p:)
+        if(present(stat)) stat = err
+    end function
+    #:endfor
+
+    elemental function to_float(s,mold) result(r)
+        ! -- In/out Variables
+        character(*), intent(in) :: s !> input string
+        real(sp), intent(in) :: mold !> dummy argument to disambiguate at compile time the generic interface
+        real(sp) :: r !> Output real value
+        ! -- Internal Variables
+        integer(1) :: p !> position within the number
+        integer(1) :: stat ! error status
+        !----------------------------------------------
+        call to_num_base(s,r,p,stat)
+    end function
+    
+    function to_float_p(s,mold,stat) result(r)
```

Comment:
the functions `to_float`, `to_double` and `to_quad`, and similarly the `to_*_p` could be simplified to one single copy in a fypp preprocessor

Reply:
Did a commit to reformat this, but now I have to include an implementation for xqp which I did not foresaw

Reply:
By any chance do you know how could I remove xdp from this loop ? :
`#:for k1, t1 in (INT_KINDS_TYPES + REAL_KINDS_TYPES)`
I could also duplicate the implementation for qp, changing wp = xdp ... but for intel compilers, these two are the same, so this would create a conflict at the generic interface.

Reply:
For the moment opted for avoiding pre-processing xdp with:
```Fortran
#:for k1, t1 in (INT_KINDS_TYPES + REAL_KINDS_TYPES)
#:if k1 != "xdp"
```

Reply:
In the `config/CMakeLists.txt` there is a test to ensure that `xdp` is not used if `qp` is defined to true.

Reply:
Note that the same "issue" appears for all `stdlib` modules that use such a `fypp` loop (e.g., in `stdlib_stats_mean.fypp`). So it should not be a problem.

Reply:
> For the moment opted for avoiding pre-processing xdp with:
> 
> ```fortran
> #:for k1, t1 in (INT_KINDS_TYPES + REAL_KINDS_TYPES)
> #:if k1 != "xdp"
> ```

If you want to avoid `xdp` for all compilers, then I think it is the best option currently.

Reply:
So from here
>In the config/CMakeLists.txt there is a test to ensure that xdp is not used if qp is defined to true.

I should understand that stdlib shall always be built with either xdp or qp, but never both, right? In that case another option would be to rename the qp implementation an "recycle it"

Reply:
If the compiler supports both `xdp` and `qp`, then both versions of the procedure will be generated. However, they will have differnet names with the suffix `xdp` or `qp`.

I am not sure I understand well your issue. Is it because there is no implementation for `xdp`?

Reply:
My question is wether we should guarantee simultaneous support for both xdp and qp ? with gfortran these two have different identifiers (xdp = 10, qp=16) with intel both are (xdp = 16, qp=16). For the latter this creates an issue as for the interfaces it can not distinguish between both. (In the case WITH_QP and WITH_XDP are both set to True)

Reply:
> My question is wether we should guarantee simultaneous support for both xdp and qp ? 

I am in favour to support both simulataneously, as for other procedures in `stdlib`.

> with gfortran these two have different identifiers (xdp = 10, qp=16) with intel both are (xdp = 16, qp=16). 

> For the latter this creates an issue as for the interfaces it can not distinguish between both. (In the case WITH_QP and WITH_XDP are both set to True)

there should be no issue with Intel when using `cmake`, because `WITH_XDP` is set to `false` in the config/CmakeList.txt. Or do I miss something?

Reply:
OK, just added support for xdp based on the qp implementation.